packages/fuels-contract/src/execution_script.rs

```diff
@@ -107,10 +111,59 @@ impl ExecutableFuelCall {
             .iter()
             .any(|r|
                 matches!(r, Receipt::ScriptResult { result, .. } if *result != ScriptExecutionResult::Success)
-        ) {
+            ) {
             return Err(Error::RevertTransactionError(Default::default(), receipts));
         }
 
         Ok(receipts)
     }
+
+    pub fn extend_transaction_inputs(&mut self, inputs: Vec<Input>) -> Result<(), Error> {
+        let (tx_utxo_ids, tx_message_ids) = self.extract_inputs_from_tx();
+
+        return match Self::check_inputs_for_duplicates(&inputs, &tx_utxo_ids, &tx_message_ids) {
+            Some(input) => {
+                Err(Error::InvalidData(
+                    format!(
+                        "{} {}",
+                        "The input with utxo_id or message_id that you are trying to add already exists in transaction inputs: ",
+                        CheckError::DuplicateInputUtxoId { utxo_id: *input.utxo_id().unwrap() }
+                    )))
+            }
+            None => {
+                self.tx.inputs_mut().extend(inputs);
+                Ok(())
+            }
+        };
+    }
+
+    fn extract_inputs_from_tx(&mut self) -> (HashSet<Bytes32>, HashSet<MessageId>) {
+        let (tx_utxo_ids, tx_message_ids): (HashSet<Bytes32>, HashSet<MessageId>) =
+            self.tx.inputs().iter().partition_map(|input| match input {
+                Input::CoinSigned { utxo_id, .. } => Either::Left(utxo_id.tx_id()),
+                Input::CoinPredicate { utxo_id, .. } => Either::Left(utxo_id.tx_id()),
+                Input::Contract { utxo_id, .. } => Either::Left(utxo_id.tx_id()),
+                Input::MessageSigned { message_id, .. } => Either::Right(message_id),
+                Input::MessagePredicate { message_id, .. } => Either::Right(message_id),
+            });
+        (tx_utxo_ids, tx_message_ids)
+    }
+
+    fn check_inputs_for_duplicates(
+        inputs: &[Input],
+        tx_utxo_ids: &HashSet<Bytes32>,
+        tx_message_ids: &HashSet<MessageId>,
+    ) -> Option<Input> {
+        inputs.to_owned().iter().cloned().find(|input| match input {
+            Input::CoinSigned { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
+            Input::CoinPredicate { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
+            Input::Contract { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
+            Input::MessageSigned { message_id, .. } => tx_message_ids.contains(message_id),
+            Input::MessagePredicate { message_id, .. } => tx_message_ids.contains(message_id),
+        })
+    }
```

Comment:
```suggestion
    fn check_inputs_for_duplicates(
        inputs: &[Input],
        tx_utxo_ids: &HashSet<Bytes32>,
        tx_message_ids: &HashSet<MessageId>,
    ) -> Option<Input> {
        inputs
            .iter()
            .find(|input| match input {
                Input::CoinSigned { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
                Input::CoinPredicate { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
                Input::Contract { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
                Input::MessageSigned { message_id, .. } => tx_message_ids.contains(message_id),
                Input::MessagePredicate { message_id, .. } => tx_message_ids.contains(message_id),
            })
            .cloned()
    }
```
No need to clone the whole collection twice, find the element first and then clone only that element.

Or better still, return a `Option<&Input>` because you're only calling `utxo_id(&self)`  on the returned value. So you can do that without any copying/cloning:
```rust
    fn check_inputs_for_duplicates<'a>(
        inputs: &'a [Input],
        tx_utxo_ids: &HashSet<Bytes32>,
        tx_message_ids: &HashSet<MessageId>,
    ) -> Option<&'a Input> {
        inputs.iter().find(|input| match input {
            Input::CoinSigned { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
            Input::CoinPredicate { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
            Input::Contract { utxo_id, .. } => tx_utxo_ids.contains(utxo_id.tx_id()),
            Input::MessageSigned { message_id, .. } => tx_message_ids.contains(message_id),
            Input::MessagePredicate { message_id, .. } => tx_message_ids.contains(message_id),
        })
    }
```